File: mc-java/src/main/java/io/spine/tools/mc/java/gradle/plugins/ProtoDataConfigPlugin.java
/*
 * Copyright 2023, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.tools.mc.java.gradle.plugins;

import com.google.common.collect.ImmutableList;
import io.spine.protodata.gradle.CodegenSettings;
import io.spine.protodata.gradle.plugin.LaunchProtoData;
import io.spine.tools.gradle.Artifact;
import io.spine.tools.mc.java.gradle.McJava;
import io.spine.tools.mc.java.gradle.Validation;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.artifacts.Dependency;

import static io.spine.tools.fs.DirectoryName.kotlin;
import static io.spine.tools.mc.java.gradle.Artifacts.toolBase;
import static io.spine.tools.mc.java.gradle.Projects.getGeneratedGrpcDirName;
import static io.spine.tools.mc.java.gradle.Projects.getGeneratedJavaDirName;
import static io.spine.tools.mc.java.gradle.Projects.getMcJava;
import static java.io.File.separatorChar;
import static java.lang.String.format;
import static java.util.Objects.requireNonNullElseGet;

/**
 * The plugin that configures ProtoData for the associated project.
 *
 * <p>We use ProtoData and the Validation library to generate validation code right inside
 * the Protobuf message classes. This plugin applies the {@code io.spine.protodata} plugin,
 * configures its extension, writes the ProtoData configuration file, and adds the required
 * dependencies to the target project.
 */
final class ProtoDataConfigPlugin implements Plugin<Project> {

    private static final String PROTO_DATA_ID = "io.spine.protodata";
    private static final String CONFIG_SUBDIR = "protodata-config";

    @SuppressWarnings("DuplicateStringLiteralInspection")
        // Could be duplicated in auto-generated Gradle code via script plugins in `buildSrc`.
    private static final String PROTODATA_CONFIGURATION = "protoData";
    private static final String IMPL_CONFIGURATION = "implementation";

    /**
     * Applies the {@code io.spine.protodata} plugin to the project and, if the user needs
     * validation code generation, configures ProtoData to generate Java validation code.
     *
     * <p>ProtoData configuration is a tricky operation because of Gradle's lifecycle.
     * We need to squeeze our configuration before the {@code LaunchProtoData} task is configured.
     * This means adding the {@code afterEvaluate(..)} hook before the ProtoData Gradle plugin
     * is applied to the project.
     */
    @Override
    public void apply(Project target) {
        target.afterEvaluate(ProtoDataConfigPlugin::configureProtoData);
        target.getPluginManager()
              .apply(PROTO_DATA_ID);
    }

    private static void configureProtoData(Project target) {
        configurePlugins(target);

        var tasks = target.getTasks();
        tasks.withType(LaunchProtoData.class, task -> {
            var name = task.getName();
            var taskName = format("writeConfigFor_%s", name);
            var configTask = tasks.create(
                    taskName,
                    GenerateProtoDataConfig.class,
                    t -> linkConfigFile(target, task, t)
            );
            task.dependsOn(configTask);
        });
    }

    private static void addDependency(
            Project project,
            String configurationName,
            Artifact artifact
    ) {
        var dependency = findDependency(project, artifact);
        project.getDependencies()
               .add(configurationName, requireNonNullElseGet(dependency, artifact::notation));
    }

    private static void addDependencies(
            Project project,
            String configurationName,
            Artifact... artifacts
    ) {
        for (var artifact : artifacts) {
            addDependency(project, configurationName, artifact);
        }
    }

    /**
     * Configures ProtoData with plugins, for the given Gradle project.
     */
    private static void configurePlugins(Project project) {
        var codegen = project.getExtensions()
                             .getByType(CodegenSettings.class);

        configureValidationRendering(project, codegen);

        codegen.plugins(
                "io.spine.tools.mc.java.rejection.RejectionPlugin"
        );

        codegen.setSubDirs(ImmutableList.of(
                getGeneratedJavaDirName().value(),
                getGeneratedGrpcDirName().value(),
                kotlin.value()
        ));

        addDependencies(
                project, PROTODATA_CONFIGURATION,
                McJava.base(),
                McJava.annotation(),
                McJava.rejection(),
                toolBase()
        );
    }

    private static void configureValidationRendering(Project project, CodegenSettings codegen) {
        codegen.plugins(
                "io.spine.validation.java.JavaValidationPlugin"
        );
        var version = getMcJava(project).codegen.validation().getVersion().get();
        addDependency(project, PROTODATA_CONFIGURATION, Validation.javaCodegenBundle(version));
        addDependency(project, IMPL_CONFIGURATION, Validation.javaRuntime(version));
    }

    private static
    void linkConfigFile(Project target, LaunchProtoData task, GenerateProtoDataConfig t) {
        var targetFile = t.getTargetFile();
        var fileName = t.getName() + ".bin";
        var defaultFile = target.getLayout()
                                .getBuildDirectory()
                                .file(CONFIG_SUBDIR + separatorChar + fileName);
        targetFile.convention(defaultFile);
        task.getConfigurationFile()
            .set(targetFile);
    }

    private static @Nullable Dependency findDependency(Project project, Artifact artifact) {
        var dependencies = project.getConfigurations().stream()
                .flatMap(c -> c.getDependencies().stream());

        var found = dependencies.filter((d) ->
            d.getGroup() != null
                    && d.getGroup().equals(artifact.group())
                    && d.getName().equals(artifact.name())
        ).findFirst().orElse(null);
        return found;
    }
}

File: mc-java/src/main/kotlin/io/spine/tools/mc/java/gradle/plugins/ProtoDataConfigPlugin.kt
/*
 * Copyright 2025, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

@file:Suppress("TooManyFunctions") // Prefer smaller configuration steps as `fun` over the limit.

package io.spine.tools.mc.java.gradle.plugins

import io.spine.protodata.gradle.Names
import io.spine.protodata.gradle.Names.GRADLE_PLUGIN_ID
import io.spine.protodata.gradle.plugin.LaunchProtoData
import io.spine.protodata.java.style.JavaCodeStyleFormatterPlugin
import io.spine.protodata.gradle.plugin.protoDataWorkingDir
import io.spine.protodata.params.WorkingDirectory
import io.spine.tools.fs.DirectoryName
import io.spine.tools.gradle.Artifact
import io.spine.tools.mc.annotation.ApiAnnotationsPlugin
import io.spine.tools.mc.java.comparable.ComparablePlugin
import io.spine.tools.mc.java.entity.EntityPlugin
import io.spine.tools.mc.java.gradle.McJava.allPlugins
import io.spine.tools.mc.java.gradle.ValidationSdk
import io.spine.tools.mc.java.gradle.generatedGrpcDirName
import io.spine.tools.mc.java.gradle.generatedJavaDirName
import io.spine.tools.mc.java.gradle.mcJava
import io.spine.tools.mc.java.gradle.plugins.ProtoDataConfigPlugin.Companion.VALIDATION_PLUGIN_CLASS
import io.spine.tools.mc.java.gradle.plugins.ProtoDataConfigPlugin.Companion.WRITE_PROTODATA_SETTINGS
import io.spine.tools.mc.java.gradle.settings.CodegenSettings
import io.spine.tools.mc.java.marker.MarkerPlugin
import io.spine.tools.mc.java.mgroup.MessageGroupPlugin
import io.spine.tools.mc.java.signal.SignalPlugin
import io.spine.tools.mc.java.signal.rejection.RThrowablePlugin
import io.spine.tools.mc.java.uuid.UuidPlugin
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.artifacts.Dependency
import org.gradle.api.provider.Provider
import org.gradle.kotlin.dsl.getByType
import org.gradle.kotlin.dsl.register
import org.gradle.kotlin.dsl.withType
import io.spine.protodata.gradle.CodegenSettings as ProtoDataSettings
import io.spine.protodata.plugin.Plugin as ProtoDataPlugin

/**
 * The plugin that configures ProtoData for the associated project.
 *
 * This plugin does the following:
 *   1. Applies the `io.spine.protodata` Gradle plugin to the project.
 *   2. Configures the ProtoData extension of a Gradle project, passing codegen
 *      plugins of ProtoData, such
 *      as [JavaValidationPlugin][io.spine.validation.java.JavaValidationPlugin].
 *   3. Creates a [WriteProtoDataSettings] task for passing configuration to ProtoData, and
 *      links it to the [LaunchProtoData] task.
 *   4. Adds required dependencies.
 */
internal class ProtoDataConfigPlugin : Plugin<Project> {

    /**
     * Applies the `io.spine.protodata` plugin to the project and, if the user needs
     * validation code generation, configures ProtoData to generate Java validation code.
     *
     * ProtoData configuration is a tricky operation because of Gradle's lifecycle.
     * We need to squeeze our configuration before the `LaunchProtoData` task is configured.
     * This means adding the `afterEvaluate(..)` hook before the ProtoData Gradle plugin
     * is applied to the project.
     */
    override fun apply(project: Project) {
        project.afterEvaluate {
            it.configureProtoData()
        }
        project.pluginManager.apply(GRADLE_PLUGIN_ID)
    }

    companion object {

        /**
         * The name of the task for writing ProtoData settings.
         */
        const val WRITE_PROTODATA_SETTINGS = "writeProtoDataSettings"

        /**
         * The name of the Validation plugin for ProtoData.
         */
        const val VALIDATION_PLUGIN_CLASS = "io.spine.validation.java.JavaValidationPlugin"
    }
}

private fun Project.configureProtoData() {
    configureProtoDataPlugins()
    val writeSettingsTask = createWriteSettingsTask()
    tasks.withType<LaunchProtoData>().all { task ->
        task.apply {
            dependsOn(writeSettingsTask)
            setStandardOutput(System.out)
            setErrorOutput(System.err)
        }
    }
}

private fun Project.createWriteSettingsTask(): Provider<WriteProtoDataSettings> {
    val result = tasks.register<WriteProtoDataSettings>(WRITE_PROTODATA_SETTINGS) {
        val workingDir = WorkingDirectory(protoDataWorkingDir.asFile.toPath())
        val settingsDir = workingDir.settingsDirectory.path.toFile()
        val settingsDirProvider = project.layout.dir(provider { settingsDir })
        this.settingsDir.set(settingsDirProvider)
    }
    return result
}

/**
 * Configures ProtoData with plugins for the given Gradle project.
 */
private fun Project.configureProtoDataPlugins() {
    // Pass the uber JAR of McJava so that plugins from all the submodules are available.
    addUserClasspathDependency(allPlugins)

    val protodata = extensions.getByType<ProtoDataSettings>()
    protodata.setSubdirectories()

    configureValidation(protodata)
    configureSignals(protodata)

    protodata.run {
        addPlugin<MarkerPlugin>()
        addPlugin<MessageGroupPlugin>()
        addPlugin<UuidPlugin>()
        addPlugin<ComparablePlugin>()
        addPlugin<EntityPlugin>()

        // Annotations should follow `SignalPlugin` and `EntityPlugin`
        // so that their output is annotated too.
        addPlugin<ApiAnnotationsPlugin>()

        // The Java style formatting comes last to conclude all the rendering.
        addPlugin<JavaCodeStyleFormatterPlugin>()
    }
}

private val Project.messageOptions: CodegenSettings
    get() = mcJava.codegen!!

private fun ProtoDataSettings.setSubdirectories() {
    subDirs = listOf(
        generatedJavaDirName.value(),
        generatedGrpcDirName.value(),
        DirectoryName.kotlin.value()
    )
}

private fun Project.configureValidation(protodata: ProtoDataSettings) {
    val validationConfig = messageOptions.validation
    val version = validationConfig.version.get()
    if (validationConfig.enabled.get()) {
        addUserClasspathDependency(ValidationSdk.javaCodegenBundle(version))
        protodata.plugins(
            VALIDATION_PLUGIN_CLASS
        )
    } else {
        addUserClasspathDependency(ValidationSdk.configuration(version))
    }

    // We add the dependency on runtime anyway for the following reasons:
    //  1. We do not want users to change their Gradle build files when they turn on or off
    //     code generation for the validation code.
    //
    //  2. We have run-time validation rules that are going to be used in parallel with
    //     the generated code. This includes current and new implementation for validation
    //     rules for the already existing generated Protobuf code.
    //
    addDependency("implementation", ValidationSdk.javaRuntime(version))
}

private fun Project.configureSignals(protodata: ProtoDataSettings) {
    protodata.addPlugin<SignalPlugin>()

    val rejectionCodegen = messageOptions.rejections
    if (rejectionCodegen.enabled.get()) {
        protodata.addPlugin<RThrowablePlugin>()
    }
}

private fun Project.addUserClasspathDependency(vararg artifacts: Artifact) = artifacts.forEach {
    addDependency(Names.USER_CLASSPATH_CONFIGURATION, it)
}

private fun Project.addDependency(configuration: String, artifact: Artifact) {
    val dependency = findDependency(artifact) ?: artifact.notation()
    dependencies.add(configuration, dependency)
}

private fun Project.findDependency(artifact: Artifact): Dependency? {
    val dependencies = configurations.flatMap { c -> c.dependencies }
    val found = dependencies.firstOrNull { d ->
        artifact.group() == d.group // `d.group` could be `null`.
                && artifact.name() == d.name
    }
    return found
}

private inline fun <reified T : ProtoDataPlugin> ProtoDataSettings.addPlugin() {
    plugins(T::class.java.name)
}
